### ISSUE TITLE
gdal2tiles.py tile borders error

### ISSUE THREAD

**Issue by Gigas002** [OSGeo/gdal#1087]
## Expected behavior and actual behavior.

I wanted to work with gdal2tiles.py like I work with MapTiler Pro console version:
`maptiler “file” -o “directory” -geodetic -resampling bilinear -srs "+proj=longlat +datum=WGS84 +no_defs" -nodata 0 0 0 -bg 0 0 0 -f png32 -tms -zoom 0 14`

So the parameters for my gdal2tiles.py were:
`gdal2tiles "file" "directory" -p geodetic -r bilinear -s "+proj=longlat +datum=WGS84 +no_defs" -a 0.0 --tmscompatible -z 10-14`

The structure of output data is fine, but the quality of gdal2tile.py's tiles are worse: they're more pixelated (can be fixed using `-r cubicspline` parameter), but more importantly the borders of tiles don't fit each other very well if you'll zoom close enough (maptiler's borders are MUCH better, almost invisible at that zooming). I attached the gdal2tiles.py screenshot of these weird borders:
![screenshot](https://user-images.githubusercontent.com/24297712/48281747-9b6ace80-e468-11e8-9db5-40ac151f12da.png)

I'd have attached the input GTiff too, but github doesn't let me attach >10mb files (the archive is ~50mbs).

There are also a small issue with overall result's border when using `-p geodetic` if input GTiff has black borders (they're not replaced by transparacy, but replaced in `-p mercator` mode), but that can be fixed by pre-gdalwarping the input GTiff.

Tried with both python 2.7 and 3.7 x64. I also tried to do the same with latest releases of QGIS 2 and 3, but the results are the same.

## Steps to reproduce the problem.

`python pgdal2tiles.py "file" "directory" -p geodetic -r bilinear -s "+proj=longlat +datum=WGS84 +no_defs" -a 0.0 --tmscompatible -z 10-14 --config GDAL_DATA D:/gdal/bin/gdal-data`

## Operating system

Win10 Pro 1809 x64/Win8.1 x64

## GDAL version and provenance

All releases was taken as complete binary packages from [GisInternals](http://www.gisinternals.com/).
x64 2.2.3 stable release;
x64 2.3.0 stable release;
x64 2.3.2 stable release;
x64 latest daily dev version;

**Comment by jratike80** [comment]
@Gigas002, could you tell why did you close the issue?

**Comment by Gigas002** [comment]
@jratike80 Sorry, I closed it for others not to spend a lot of time on it.
I was preparing the answer, I think I've found the place, where the error occurs. Probably it's the `geo_query` function, I suppose there are some values lost because of type convertions. 
I've rewritten needed part of [gdal2tiles.py](https://github.com/OSGeo/gdal/blob/master/gdal/swig/python/scripts/gdal2tiles.py) script on c# and located that place after several tests.
I've rewritten the mentioned function and replaced with c#'s methods (using classes from `System.Drawing` and `System.Drawing.Imaging` namespaces) to create tiles instead of gdal's `ReadRaster` and `WriteRaster`, [here's](https://github.com/Gigas002/Gdal2Tiles/blob/master/Gdal2Tiles/GTiff2Tiles.cs) class that I made. Just before the value of `GeoQuery` method returns, I calculate the shift value - the difference between actual positions (double, represent pixels) and converted to int values, after what I add these values to array's sizes.
That borders issue didn't solved completely, yet it looks much better, the same as MapTiler's output.
